base/compiler/ssair/tarjan.jl

```diff
@@ -0,0 +1,299 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+using Core.Compiler: DomTree, CFG, BasicBlock, StmtRange, dominates
+
+struct SCCStackItem
+    v::Int32
+    # which child of `v` to scan
+    child::Int32
+    # the location of `parent` in the stack
+    parent::Int32
+    # the index in the (pre-order traversal of the) DFS tree
+    preorder::Int32
+    # the minimum node (by pre-order index) reachable from any node in the DFS sub-tree rooted at `v`
+    minpreorder::Int32
+    # whether this node is reachable from BasicBlock #1
+    live::Bool
+end
+
+function SCCStackItem(item::SCCStackItem; child=item.child,
+                      minpreorder=item.minpreorder, live=item.live)
+    return SCCStackItem(
+        item.v,        # v
+        child,         # child
+        item.parent,   # parent
+        item.preorder, # preorder
+        minpreorder,   # minpreorder
+        live,          # live
+    )
+end
+
+struct CFGReachability
+    irreducible::BitVector # BBNumber -> Bool
+    scc::Vector{Int}       # BBNumber -> SCCNumber
+    domtree::DomTree
```

Comment:
What keeps this domtree in sync during edge removal?

Reply:
It's intentionally kept out of sync: maintaining the dominance tree is O(N^2) at least while this is O(N) (for reducible CFGs) so there's a performance motivation, but there's a deeper reason w.r.t. SSA legality: I am pretty sure that if you want to use the dominance tree to locate phi nodes, you actually want to keep it out of sync.

My reasoning is that after edge deletion if you update the domtree, you'll remove dominance frontiers where you still have (single-element) ϕ nodes. Eliminating those ϕ nodes requires both updating the domtree and re-ordering the basic blocks and all instructions to match, which is quite expensive.

In order words, updating the DomTree seems like it requires re-legalizing the SSA since the latter is defined in terms of the former.

Reply:
Keeping it out of sync like this means that we keep these single-element PhiNodes around and potentially don't notice some irreducible control flow that actually became reducible.

It also means that any passes reasoning about dominance directly (rather than following use-def flow) will take a hit to precision, but I'm not sure if we do that anywhere

Reply:
I think it would make me happier if rather than keeping the domtree, we just kept a list of which edges we consider the backedges. It'd probably be faster too, since dominates is not quite O(1).

Reply:
I think that sidesteps the issue, since external abstract interpreters might use the domtree downstream (as we do now). I'd prefer to align on what domtree we need to provide downstream, and make this work for that.

On the other hand, I'm still not convinced the updated domtree would be correct for many downstream uses (unless SSA is re-legalized). Consider for example:
- Inserting a new def-like: If the domtree reflects the true dominance rather than the dominance at SSA construction, the `iterated_dominance_frontier` will fail to report blocks where a single-element phi is required due to stmt ordering.
- Walking to all uses of stmt that flows through PhiNodes e.g. https://github.com/JuliaLang/julia/blob/47d31acd1494dfed4f8f361d67fe1fb8c3888656/base/compiler/optimize.jl#L903-L908 If the domtree is updated after edge deletions, this will fail to mark inconsistent single-element PhiNodes that propagate results from the GotoIfNot.

Reply:
Also it's generally sound to under-approximate dominance relations like this - since it's equivalent to assuming the presence of control-flow edges that don't get exercised at runtime.

I do agree that _insertions_ of new control flow edges (of the same kind that require SSA re-legalization anyway) do need some kind of update to the domtree though

Reply:
I was wondering why we haven't run into this already, and it turns out this is an existing bug: https://github.com/JuliaLang/julia/issues/53013